iris_grib/_save_rules.py

```diff
@@ -649,6 +650,65 @@ def grid_definition_template_30(cube, grib):
     eccodes.codes_set(grib, "longitudeOfSouthernPole", 0)
 
 
+def grid_definition_template_140(cube, grib):
+    """
+    Set keys within the provided grib message based on
+    Grid Definition Template 3.140.
+
+    Template 3.140 is used to represent a Lambert Azimuthual Equal Area grid.
+
+    """
+
+    eccodes.codes_set(grib, "gridDefinitionTemplateNumber", 140)
+
+    # Retrieve some information from the cube.
+    y_coord = cube.coord(dimensions=[0])
+    x_coord = cube.coord(dimensions=[1])
+    cs = y_coord.coord_system
+
+    # Normalise the coordinate values to millimetres - the resolution
+    # used in the GRIB message.
+    y_mm = points_in_unit(y_coord, 'mm')
+    x_mm = points_in_unit(x_coord, 'mm')
+
+    # Encode the horizontal points.
+
+    # NB. Since we're already in millimetres, our tolerance for
+    # discrepancy in the differences is 1.
+    try:
+        x_step = step(x_mm, atol=1)
+        y_step = step(y_mm, atol=1)
+    except ValueError:
+        msg = ('Irregular coordinates not supported for Lambert '
+               'Azimuthal Equal Area.')
+        raise TranslationError(msg)
+    eccodes.codes_set(grib, 'Dx', abs(x_step))
+    eccodes.codes_set(grib, 'Dy', abs(y_step))
+
+    horizontal_grid_common(cube, grib, xy=True)
+
+    # Transform first point into geographic CS
+    geog = cs.ellipsoid if cs.ellipsoid is not None else GeogCS(1)
+    first_x, first_y = geog.as_cartopy_crs().transform_point(
+        x_coord.points[0],
+        y_coord.points[0],
+        cs.as_cartopy_crs())
+    first_x = first_x % 360
+    central_lon = cs.longitude_of_projection_origin % 360
+    central_lat = cs.latitude_of_projection_origin
+
+    eccodes.codes_set(grib, "latitudeOfFirstGridPoint",
+                      int(np.round(first_y / _DEFAULT_DEGREES_UNITS)))
+    eccodes.codes_set(grib, "longitudeOfFirstGridPoint",
+                      int(np.round(first_x / _DEFAULT_DEGREES_UNITS)))
+    eccodes.codes_set(grib, 'standardParallel',
+                      central_lat / _DEFAULT_DEGREES_UNITS)
+    eccodes.codes_set(grib, 'centralLongitude',
+                      central_lon / _DEFAULT_DEGREES_UNITS)
+    eccodes.codes_set(grib, 'resolutionAndComponentFlags',
+                      0x1 << _RESOLUTION_AND_COMPONENTS_GRID_WINDS_BIT)
```

Comment:
I guessed based on the other code, particularly that of loading.  I'm not a GRIB expert - I was hoping this would be checked as part of the review process!

Reply:
OK, sorry for putting the onus on you. Most Iris-GRIB PRs I have reviewed have been from GRIB experts by coincidence so I have developed a bad habit.

Here's what I have after further investigation. **TLDR you've already covered all the bases**:

- Octets 15 to 30: handled by this line:
  https://github.com/SciTools/iris-grib/blob/70d35ef0073f27c22345be07d3eca0301c6873e5/iris_grib/_save_rules.py#L688
- Octets 31 to 38: `numberOfPointsAlongXAxis` and `numberOfPointsAlongYAxis` are aliased to `Nx` and `Ny`, which are handled by this line:
  https://github.com/SciTools/iris-grib/blob/70d35ef0073f27c22345be07d3eca0301c6873e5/iris_grib/_save_rules.py#L688
- Octets 39 to 55: you are already handling
- Octets 56 to 63: `xDirectionGridLengthInMillimetres` and `yDirectionGridLengthInMillimetres` are alised to `Dx` and `Dy`, which are handled by these lines:
  https://github.com/SciTools/iris-grib/blob/70d35ef0073f27c22345be07d3eca0301c6873e5/iris_grib/_save_rules.py#L685-L686
- Octet 64: handled by this line:
  https://github.com/SciTools/iris-grib/blob/70d35ef0073f27c22345be07d3eca0301c6873e5/iris_grib/_save_rules.py#L688